packages/core/src/loaders/progressiveLoader.ts

```diff
@@ -0,0 +1,382 @@
+      if (done) {
+        updateStageStatus(stageStatus, request.stage, reason);
+      }
+    };
+    let loadedPromise;
+    if (options.target?.arrayBuffer) {
+      loadedPromise = loadAndCacheImage(imageId, options);
+    } else {
+      loadedPromise = loadImage(imageId, options);
+    }
+    const uncompressedIterator = ProgressiveIterator.as(loadedPromise);
+    let complete = false;
+
+    uncompressedIterator
+      .forEach(async (image, done) => {
+        const oldStatus = imageStatus[imageId];
+        if (!image) {
+          console.warn('No image retrieved', imageId);
+          return;
+        }
+        const { status } = image;
+        complete ||= status === ImageStatus.DONE;
+        if (oldStatus !== undefined && oldStatus > status) {
+          // We already have a better status, so don't update it
+          updateStageStatus(stageStatus, request.stage, null, true);
+          return;
+        }
+        imageStatus[imageId] = ImageStatus.LOADING;
+
+        listener.successCallback(imageId, image, status);
+        imageStatus[imageId] = status;
+        displayedIterator.add(image);
+        if (done) {
+          updateStageStatus(stageStatus, request.stage);
+        }
+        fillNearbyFrames(listener, imageStatus, request, image, options);
+      }, errorCallback)
+      .finally(() => {
+        if (!complete && next) {
+          if (cache.getImageLoadObject(imageId)) {
+            cache.removeImageLoadObject(imageId);
+          }
+          addRequest(next, options.streamingData);
+        } else {
+          outstandingRequests--;
+          for (let skip = next; skip; skip = skip.next) {
+            updateStageStatus(stageStatus, skip.stage, null, true);
+          }
+        }
+        if (outstandingRequests <= 0) {
+          displayedIterator.resolve();
+        }
+      });
+    const doneLoad = uncompressedIterator.getDonePromise();
+    // Errors already handled above in the callback
+    return doneLoad.catch((e) => null);
+  }
+
+  /** Adds a rquest to the image load pool manager */
+  function addRequest(request, streamingData = {}) {
+    const { imageId, stage } = request;
+    const baseOptions = listener.getTargetOptions(imageId);
+    if (!baseOptions) {
+      // Image no longer of interest
+      return;
+    }
+    const options = {
+      ...baseOptions,
+      retrieveType: stage.retrieveType,
+      streamingData,
+    };
+    const priority = stage.priority ?? -5;
+    const requestType = stage.requestType || RequestType.Interaction;
+    const additionalDetails = { imageId };
+
+    imageLoadPoolManager.addRequest(
+      sendRequest.bind(this, request, options),
+      requestType,
+      additionalDetails,
+      priority
+    );
+  }
+
+  // The actual function is to just setup the interleave and add the
+  // requests, with all the actual work being handled by the nested functions
+  const interleaved = interleave(imageIds, retrieveOptions, stageStatus);
```

Comment:
This should be refactored so that any custom `retrieveOptions` or custom ordering function people have can be applied here. Instead of hardcoding the interleave, the function name is simply `load`.

Reply:
I'm quite surprised that I see interleave here even for loadSingle, I don't know how this works but most likely this should be the retrieveConfiguration callback not hardcoded interleave